lib/impl/AdobeStateStore.js

```diff
@@ -0,0 +1,377 @@
+
+const ADOBE_STATE_STORE_ENDPOINT = {
+  [PROD_ENV]: 'http://localhost:8080', // TODO:
+  [STAGE_ENV]: 'http://localhost:8080' // TODO:
+}
+
+/* *********************************** typedefs *********************************** */
+
+/**
+ * AdobeStateStoreCredentials
+ *
+ * @typedef AdobeStateStoreCredentials
+ * @type {object}
+ * @property {string} namespace the state store namespace
+ * @property {string} apikey the state store api key
+ */
+
+/**
+ * AdobeStateStore put options
+ *
+ * @typedef AdobeStateStorePutOptions
+ * @type {object}
+ * @property {number} ttl time-to-live for key-value pair in seconds, defaults to 24 hours (86400s). Set to < 0 for no expiry. A
+ * value of 0 sets default.
+ */
+
+/**
+ * AdobeStateStore get return object
+ *
+ * @typedef AdobeStateStoreGetReturnValue
+ * @type {object}
+ * @property {string|null} expiration ISO date string of expiration time for the key-value pair, if the ttl is infinite
+ * expiration=null
+ * @property {any} value the value set by put
+ */
+
+/* *********************************** helpers *********************************** */
+
+// eslint-disable-next-line jsdoc/require-jsdoc
+async function _wrap (promise, params) {
+  let response
+  try {
+    response = await promise
+    // reuse code in exception handler, for any other network exceptions
+    if (!response.ok) {
+      // no exception on 404
+      if (response.statusCode === 404) {
+        return null
+      } else {
+        const e = new Error(response.statusText)
+        e.statusCode = response.statusCode
+        e.internal = response
+        throw e
+      }
+    }
+  } catch (e) {
+    const statusCode = e.statusCode || e.code
+    const copyParams = cloneDeep(params)
+    logger.debug(`got internal error with status ${statusCode}: ${e.message} `)
+    switch (statusCode) {
+      case 401:
+        return logAndThrow(new codes.ERROR_UNAUTHORIZED({ messageValues: ['underlying DB provider'], sdkDetails: copyParams }))
+      case 403:
+        return logAndThrow(new codes.ERROR_BAD_CREDENTIALS({ messageValues: ['underlying DB provider'], sdkDetails: copyParams }))
+      case 429:
+        return logAndThrow(new codes.ERROR_REQUEST_RATE_TOO_HIGH({ sdkDetails: copyParams }))
+      default:
+        return logAndThrow(new codes.ERROR_INTERNAL({ messageValues: [`unexpected response from provider with status: ${statusCode}`], sdkDetails: { ...cloneDeep(params), _internal: e.internal } }))
+    }
+  }
+  return response
+}
+
+// eslint-disable-next-line jsdoc/require-jsdoc
+function validateInput (input, schema, details) {
+  const validation = schema.validate(input)
+  if (validation.error) {
+    logAndThrow(new codes.ERROR_BAD_ARGUMENT({
+      messageValues: [validation.error.message],
+      sdkDetails: cloneDeep(details)
+    }))
+  }
+}
+
+// eslint-disable-next-line jsdoc/require-jsdoc
+function validateKey (key, details, label = 'key') {
```

Comment:
Validation is different and more restrictive now, we also need to validate for namespaces:

```
export const namespaceRgxStr = '^(development-)?([0-9]{3,10})-([a-z0-9]{1,20})(-([a-z0-9]{1,20}))?$'
// The regex for keys, allowed chars are alphanumerical with _ and -
export const keyRgxStr = `^[a-zA-Z0-9-_-]{1,${MAX_KEY_SIZE}}$`
```

But those checks are also on the server so we may delay the validation checks on the client for another PR.

Reply:
good idea, I'll remove the validation checks on the client

Reply:
done